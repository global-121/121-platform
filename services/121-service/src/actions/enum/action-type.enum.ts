export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
