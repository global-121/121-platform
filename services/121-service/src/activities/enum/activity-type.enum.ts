export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
