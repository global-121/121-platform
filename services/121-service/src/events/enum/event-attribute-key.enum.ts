export enum EventAttributeKeyEnum {
  reason = 'reason',
  oldValue = 'oldValue',
  newValue = 'newValue',
  fieldName = 'fieldName',
  duplicateRegistrationId = 'duplicateRegistrationId',
  duplicateRegistrationProgramId = 'duplicateRegistrationProgramId',
}
