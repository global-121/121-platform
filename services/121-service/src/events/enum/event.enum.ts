export enum EventEnum {
  registrationDataChange = 'registrationDataChange',
  financialServiceProviderChange = 'financialServiceProviderChange',
  registrationStatusChange = 'registrationStatusChange',
  ignoredDuplication = 'ignoredDuplication',
}
