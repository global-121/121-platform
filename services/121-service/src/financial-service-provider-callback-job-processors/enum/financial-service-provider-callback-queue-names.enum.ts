export enum FinancialServiceProviderCallbackQueueNames {
  safaricomTransferCallback = 'safaricomTransferCallback',
  safaricomTimeoutCallback = 'safaricomTimeoutCallback',
}
