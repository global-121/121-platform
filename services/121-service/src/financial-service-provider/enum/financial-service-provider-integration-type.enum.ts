export enum FinancialServiceProviderIntegrationType {
  api = 'api',
  csv = 'csv',
  xml = 'xml',
}
