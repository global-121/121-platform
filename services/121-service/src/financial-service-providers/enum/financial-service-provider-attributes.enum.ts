export enum FinancialServiceProviderAttributes {
  phoneNumber = 'phoneNumber',
  nationalId = 'nationalId',
  fullName = 'fullName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
  bankAccountNumber = 'bankAccountNumber',
  whatsappPhoneNumber = 'whatsappPhoneNumber',
}
