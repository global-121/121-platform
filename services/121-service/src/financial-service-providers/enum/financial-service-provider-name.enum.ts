export enum FinancialServiceProviderName {
  intersolveVoucherWhatsapp = 'Intersolve-voucher-whatsapp',
  intersolveVoucherPaper = 'Intersolve-voucher-paper',
  intersolveVisa = 'Intersolve-visa',
  intersolveJumboPhysical = 'Intersolve-jumbo-physical',
  africasTalking = 'Africas-talking',
  belcash = 'BelCash',
  vodacash = 'VodaCash',
  bobFinance = 'BoB-finance',
  ukrPoshta = 'UkrPoshta',
  safaricom = 'Safaricom',
  commercialBankEthiopia = 'Commercial-bank-ethiopia',
  excel = 'Excel',
}

export enum FinancialServiceProviderConfigurationEnum {
  password = 'password',
  username = 'username',
  columnsToExport = 'columnsToExport',
  columnToMatch = 'columnToMatch',
  brandCode = 'brandCode',
  displayName = 'displayName',
  coverLetterCode = 'coverLetterCode',
}

export const FinancialServiceProviderConfigurationMapping: {
  [key in FinancialServiceProviderName]?: any;
} = {
  [FinancialServiceProviderName.intersolveVoucherWhatsapp]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVoucherPaper]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVisa]: [
    FinancialServiceProviderConfigurationEnum.brandCode,
    FinancialServiceProviderConfigurationEnum.coverLetterCode,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveJumboPhysical]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.africasTalking]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.belcash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.vodacash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.bobFinance]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.ukrPoshta]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.safaricom]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.commercialBankEthiopia]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.excel]: [
    FinancialServiceProviderConfigurationEnum.columnsToExport,
    FinancialServiceProviderConfigurationEnum.columnToMatch,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
};
