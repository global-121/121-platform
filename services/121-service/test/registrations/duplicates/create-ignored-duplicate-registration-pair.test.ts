import { ActivityTypeEnum } from '@121-service/src/activities/enum/activity-type.enum';
import { SeedScript } from '@121-service/src/scripts/enum/seed-script.enum';
import {
  getActivities,
  getRegistrationIdByReferenceId,
  ignoreDuplicates,
  importRegistrations,
} from '@121-service/test/helpers/registration.helper';
import {
  getAccessToken,
  resetDB,
} from '@121-service/test/helpers/utility.helper';
import {
  registrationPV5,
  registrationPV6,
} from '@121-service/test/registrations/pagination/pagination-data';
const registration1 = { ...registrationPV5 };
const registration2 = { ...registrationPV6 };
const programId = 2;
describe('Get duplicate status of registrations', () => {
  let accessToken: string;

  beforeEach(async () => {
    await resetDB(SeedScript.nlrcMultiple);
    accessToken = await getAccessToken();
    registration1.phoneNumber = '1234567890';
    registration2.phoneNumber = '1234567890';
    await importRegistrations(
      programId,
      [registration1, registration2],
      accessToken,
    );
  });

  it(`should sucessfully create an ignored duplicate registration pair and related activities`, async () => {
    const reason = 'test reason';
    const ignoreDuplicatesResult = await ignoreDuplicates({
      referenceId1: registration1.referenceId,
      referenceId2: registration2.referenceId,
      programId,
      accessToken,
      reason,
    });

    const registrationId1 = await getRegistrationIdByReferenceId({
      referenceId: registration1.referenceId,
      programId,
      accessToken,
    });
    const registrationId2 = await getRegistrationIdByReferenceId({
      referenceId: registration2.referenceId,
      programId,
      accessToken,
    });

    const activities1 = (
      await getActivities({
        programId,
        registrationId: registrationId1,
        accessToken,
      })
    ).body.data;
    const ignoreDuplicateActivity1 = activities1.find(
      (activity) => activity.type === ActivityTypeEnum.IgnoredDuplication,
    );

    const activities2 = (
      await getActivities({
        programId,
        registrationId: registrationId2,
        accessToken,
      })
    ).body.data;
    const ignoreDuplicateActivity2 = activities2.find(
      (activity) => activity.type === ActivityTypeEnum.IgnoredDuplication,
    );

    expect(ignoreDuplicatesResult.status).toBe(201);

    expect(ignoreDuplicateActivity1).toMatchObject({
      type: ActivityTypeEnum.IgnoredDuplication,
      attributes: {
        duplicateRegistrationId: String(registrationId2),
        duplicateRegistrationProgramId: expect.any(String),
        reason,
      },
      created: expect.any(String),
      id: expect.any(String),
      user: {
        id: expect.any(Number),
        username: process.env.USERCONFIG_121_SERVICE_EMAIL_ADMIN,
      },
    });
    expect(ignoreDuplicateActivity2).toMatchObject({
      type: ActivityTypeEnum.IgnoredDuplication,
      attributes: {
        duplicateRegistrationId: String(registrationId1),
        duplicateRegistrationProgramId: expect.any(String),
        reason,
      },
      created: expect.any(String),
      id: expect.any(String),
      user: {
        id: expect.any(Number),
        username: process.env.USERCONFIG_121_SERVICE_EMAIL_ADMIN,
      },
    });
  });

  it(`should fail to create an ignored duplicate registration pair if the pair already exists`, async () => {
    await ignoreDuplicates({
      referenceId1: registration1.referenceId,
      referenceId2: registration2.referenceId,
      programId,
      accessToken,
    });
    const result = await ignoreDuplicates({
      referenceId1: registration1.referenceId,
      referenceId2: registration2.referenceId,
      programId,
      accessToken,
    });
    expect(result.status).toBe(400);
    expect(result.body.message).toMatchSnapshot();

    const resultReversed = await ignoreDuplicates({
      referenceId1: registration2.referenceId,
      referenceId2: registration1.referenceId,
      programId,
      accessToken,
    });
    expect(resultReversed.status).toBe(400);
    expect(resultReversed.body.message).toMatchSnapshot();
  });

  it(`should fail to create an ignored duplicate registration pair for itself`, async () => {
    const result = await ignoreDuplicates({
      referenceId1: registration1.referenceId,
      referenceId2: registration1.referenceId,
      programId,
      accessToken,
    });
    expect(result.status).toBe(400);
    expect(result.body.message).toMatchSnapshot();
  });
});
