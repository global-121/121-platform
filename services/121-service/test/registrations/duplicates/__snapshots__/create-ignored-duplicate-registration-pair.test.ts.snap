// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get duplicate status of registrations should fail to create an ignored duplicate registration pair for itself 1`] = `"ReferenceIds are the same"`;

exports[`Get duplicate status of registrations should fail to create an ignored duplicate registration pair if the pair already exists 1`] = `"This registration pair is already ignored"`;

exports[`Get duplicate status of registrations should fail to create an ignored duplicate registration pair if the pair already exists 2`] = `"This registration pair is already ignored"`;
